### ISSUE TITLE
[AWS Autoscaler] create_node hangs forever when insufficient AWS Capacity

### ISSUE THREAD

**Issue by AdamGleave** [ray-project/ray#2138]
In aws/node_provider.py:create_node, the call to ```self.ec2.create_instances``` can block for over half an hour, stopping all other Monitor activity. This has happened to me for spot requests when AWS has no capacity, but it could plausibly happen in other situations too.

A quick fix might be to reduce the number of retries that boto makes. They support a [global parameter](http://botocore.readthedocs.io/en/latest/reference/config.html), but it doesn't seem like it can be set per request.

A more robust approach would be to use request_spot_instances, reserving create_instance only for on-demand instances; however, this would complicate the code significantly, as I think we'd need to poll for when the instances are created.

Alternately, we could switch to using {request,modify}_spot_fleet, which supports both on-demand and spot types. I think this is closest to how Amazon intends the API to be used, but it is a bit of a mismatch with the Ray autoscaling API. The main problem is in scaling down: I think we'd need to follow [this procedure](https://stackoverflow.com/questions/45870613/terminate-specific-instance-in-aws-spot-fleet) to terminate a specific node. We'd also still need to poll to discover new nodes.

## Logs

```
StandardAutoscaler: Launching 2 new nodes
StandardAutoscaler: Error during autoscaling: {} Traceback (most recent call last):
  File "/home/ec2-user/anaconda3/envs/pirl/lib/python3.6/site-packages/ray/autoscaler/autoscaler.py", line 246, in update
    self._update()
  File "/home/ec2-user/anaconda3/envs/pirl/lib/python3.6/site-packages/ray/autoscaler/autoscaler.py", line 308, in _update
    min(self.max_concurrent_launches, target_num - len(nodes)))
  File "/home/ec2-user/anaconda3/envs/pirl/lib/python3.6/site-packages/ray/autoscaler/autoscaler.py", line 455, in launch_new_node
    count)
  File "/home/ec2-user/anaconda3/envs/pirl/lib/python3.6/site-packages/ray/autoscaler/aws/node_provider.py", line 113, in create_node
    self.ec2.create_instances(**conf)
  File "/home/ec2-user/anaconda3/envs/pirl/lib/python3.6/site-packages/boto3/resources/factory.py", line 520, in do_action
    response = action(self, *args, **kwargs)
  File "/home/ec2-user/anaconda3/envs/pirl/lib/python3.6/site-packages/boto3/resources/action.py", line 83, in __call__
    response = getattr(parent.meta.client, operation_name)(**params)
  File "/home/ec2-user/anaconda3/envs/pirl/lib/python3.6/site-packages/botocore/client.py", line 324, in _api_call
    return self._make_api_call(operation_name, kwargs)
  File "/home/ec2-user/anaconda3/envs/pirl/lib/python3.6/site-packages/botocore/client.py", line 622, in _make_api_call
    raise error_class(parsed_response, operation_name)
botocore.exceptions.ClientError: An error occurred (InsufficientInstanceCapacity) when calling the RunInstances operation (reached max retries: 12): There is no Spot capacity available that matches your request.

StandardAutoscaler [2018-05-25 19:16:35.413360]: 0/2 target nodes
 - NodeIdleSeconds: Min=0 Mean=0 Max=0
 - NumNodesConnected: 1
 - NumNodesUsed: 1.0
 - ResourceUsage: 4.0/4.0 CPU, 4.0/4.0 GPU
StandardAutoscaler: Launching 2 new nodes
StandardAutoscaler [2018-05-25 19:17:15.009775]: 2/2 target nodes
```

**Comment by AdamGleave** [comment]
I looked into this a little more. Changing the number of retries in boto seems quite messy and I'd prefer not to go down that route. I think the easiest change to make would be to use the SpotInstanceRequest, and then block for some configurable period (e.g. a minute) until we know if the request is satisfied; if not, then cancel the request. This would require relaxing the assertion in autoscaler.py:launch_new_node that checks new nodes have in fact been launched, but I can't see any harm coming from this.

I'd be happy to implement this, however I don't feel it's a great solution, so would appreciate some feedback before starting.

**Comment by robertnishihara** [comment]
Will this happen any time that I try to create more instances than my limits allow?

If you are able to help out with fixing this that'd be awesome! @ericl is the person to talk to.

**Comment by ericl** [comment]
Another approach would be to do the request in another thread, so we could still use CreateInstances. This would require some additional tracking logic for outstanding requests, but would avoid blocking the main loop.

Let me know what you think.

**Comment by AdamGleave** [comment]
@robertnishihara I think if you hit instance limits EC2 returns an error immediately rather than blocking (at least I did not encounter this problem and I have hit instance limits before). The issue can be reproduced fairly reliably by trying to create spot instances with limited capacity, e.g. a p2.16xlarge request in us-west-2b will almost certainly not be fulfilled.

@ericl I'm happy with a threading solution, I was reluctant to suggest it as I hadn't seen it used elsewhere in Ray, but a search throws up some other use cases. 

The main challenge I see with introducing concurrency here is that the autoscaler keeps trying to add nodes so long as the current number of nodes is less than the target. If the requests blocks in a separate thread (but is eventually satisfied), we could end up launching a lot more nodes than was originally intended. Options I see are to either change the API and have the autoscaler expose the target number of nodes to the node provider, or to have the AWS node provider only process one instance launch request at a time (silently dropping the others). Thoughts?

**Comment by ericl** [comment]
@AdamGleave originally I was thinking of something like the following:

1. Keep a "pending nodes" counter in the autoscaler which tracks outstanding instance creation requests.
2. When checking whether to create new nodes, compare num_running + num_pending to the target.
3. When checking whether to remove nodes, compare against just num_running.

The idea to just avoid launching more than 1 instance creation request at a time could also work though, and might be simpler, since it only involves only adding a single boolean to the autoscaler state. I would do this in the autoscaler logic itself instead of the node provider so that it will work with other providers (perhaps create a new node provider object instance for the instance launcher thread, to avoid potential thread safety issues).

**Comment by AdamGleave** [comment]
@ericl That approach sounds good. The only problem I can foresee is we won't cancel requests when we no longer need them (only terminating the node once it starts); this problem seems minor, however.

An alternative that avoids this problem (and does not require threading) would be to have create_node make a [RequestSpotInstance](https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_RequestSpotInstances.html) call, which should always return immediately. nodes() would be changed to check both the running instances and unfulfilled spot instance requests. (There's potentially a race here, but it seems like it would happen very rarely, and would not have major consequences.) {internal,external}_ip would return None while the instance request is pending.

The only change I can see this requiring to the autoscaler would having _update preferentially terminate nodes without a recorded last_used time. However, I'm not familiar with this part of the code base so I may have overlooked some other implications.

Any thoughts on pros/cons of this compared to alternative approaches? I'll try and carve out some time later this week to implement whatever we decide on.

**Comment by ericl** [comment]
I think the main advantage of RequestSpotInstance would be the built-in cancellation. However, I think this is something we can add separately, for example, by cancelling the underlying HTTP request, or running the request in another process and killing the process to kill all its outstanding requests.

Another thing is that RequestSpotInstance is very AWS-specific, and would not address the more general problem of instance creation requests hanging for other reasons. Whereas, a threading / process based solution would work for GCE as well and we can make it tolerant of hangs in the general case.

**Comment by AdamGleave** [comment]
Could you elaborate on what other hangs in ```create_node``` we might expect to encounter, this might help me understand the design trade-offs better? Note we make (blocking) API calls in many other places, e.g. each call to nodes() retrieves all the currently running EC2 instances.

If there are some other cases where this is likely to arise then I'd favor the threading solution, but otherwise I feel a single-threaded approach is cleaner. I'm also not sure Amazon's API is clever enough to interpret closing the HTTP session as corresponding to cancelling the spot request. 

@hartikainen do you know if the API call to start an instance in GCE ever blocks for a significant amount of time? The [documentation](https://cloud.google.com/compute/docs/instances/create-start-instance) doesn't make this clear. If so this is definitely an argument in favor of threading. If not then this problem feels like an AWS-specific quirk that we should fix in the AWS node provider.

**Comment by ericl** [comment]
I believe that even a normal instance request can take a while if there are insufficient instances available (e.g., I've seen this when requesting large numbers of GPU instances).

My preference here is to avoid additional AWS dependencies such as createSpotInstances, for the sake of longer-term maintainability, I'm willing to increase architectural complexity a bit to avoid these special cases.

**Comment by hartikainen** [comment]
@AdamGleave Good question. I don't know for sure how long GCE instance creation might take, but I do know that the blocking doesn't happen on the library level. [instances.insert](https://cloud.google.com/compute/docs/reference/rest/v1/instances/insert) method returns an [operation](https://cloud.google.com/deployment-manager/docs/reference/latest/operations), which we then [explicitly wait to complete](https://github.com/ray-project/ray/blob/940c1b1fd1d5b168cdbab3e63f501d94eea23b4d/python/ray/autoscaler/gcp/config.py#L51).

**Comment by AdamGleave** [comment]
Closing as fixed by #2183
